src/Chapter05/Listing05.33.CustomParameterValidation.cs

```diff
@@ -0,0 +1,60 @@
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter05.Listing05_33;
+
+#region INCLUDE
+public class Program
+{
+    #region HIGHLIGHT
+    public static int Main(string[] args)
+    #endregion HIGHLIGHT
+    {
+        #region HIGHLIGHT
+        int result = 0;
+        #endregion HIGHLIGHT
+        if(args.Length != 2 ) 
+        { 
+            // Exactly two arguments must be specified; give an error
+            Console.WriteLine(
+                "ERROR:  You must specify the "
+                + "URL and the file name");
+            Console.WriteLine(
+                "Usage: Downloader.exe <URL> <TargetFileName>");
+            result = 1;
+        }
+        else
+        {
+            DownloadSSL(args[0], args[1]);
+        }
+        return result;
+    }
+
+private static void DownloadSSL(string httpsUrl, string fileName)
+{
+    #region HIGHLIGHT
+#if !NET7_0_OR_GREATER
+    httpsUrl=httpsUrl?.Trim()??
+        throw new ArgumentNullException(nameof(httpsUrl));
+    fileName=fileName?.Trim()??
+        throw new ArgumentNullException(nameof(fileName));
+    if (fileName?.Length == 0)
+    {
+        throw new ArgumentException(
+            $"{nameof(fileName)} cannot be empty or only whitespace");
+    }
+#else
+    ArgumentException.ThrowIfNullOrEmpty(httpsUrl.TrimEnd());
+    ArgumentException.ThrowIfNullOrEmpty(fileName?.Trim());
+#endif
+    if (!httpsUrl.ToUpper().StartsWith("HTTPS"))
+    {
+        throw new ArgumentException("URL must start with 'HTTPS'.");
+    }
+#endregion HIGHLIGHT
+    HttpClient client = new();
+    byte[] response =
+        client.GetByteArrayAsync(httpsUrl).Result;
+    client.Dispose();
+    File.WriteAllBytes(fileName!, response);
```

Comment:
Do you need the null forgiveness operator here? I would have suspected the compiler would know it is non-null at this point.

Reply:
For the .NET 6.0 and earlier code - yes. :( Suggestions welcome.

Reply:
I am fine leaving the null forgiveness operator in.